modules/packages/Sort.chpl

```diff
@@ -1233,52 +1233,67 @@ module QuickSort {
     import Sort.InsertionSort;
 
     // grab obvious indices
-    const lo = start,
-          hi = end,
-          mid = lo + (hi-lo+1)/2;
-    var piv = mid;
-
-    if hi - lo < 0 { // minlen {
-      // base case -- use insertion sort
-      InsertionSort.insertionSortMoveElts(Data, comparator=comparator, lo, hi);
-      return;
-    } else if hi <= lo {
-      // nothing to sort
-      return;
-    }
+    var lo = start,
+        hi = end;
 
-    // find pivot using median-of-3 method for small arrays
-    // and a "ninther" for bigger arrays. Places the pivot in
-    // Data[lo].
-    if hi - lo < 100 {
-      piv = order3(Data, lo, mid, hi, comparator);
-    } else {
-      // assumes array size > 9 at the very least
+    // keep iterating over subproblems
+    while lo < hi {
+        var mid = lo + (hi - lo + 1) / 2;
+        var piv = mid;
 
-      // median of each group of 3
-      const medLo  = order3(Data, lo,    lo+1, lo+2,  comparator);
-      const medMid = order3(Data, mid-1, mid,  mid+1, comparator);
-      const medHi  = order3(Data, hi-2,  hi-1, hi,    comparator);
-      // median of the medians
-      piv = order3(Data, medLo, medMid, medHi, comparator);
-    }
+        if hi - lo < 0 { // minlen {
+          // base case -- use insertion sort
+          InsertionSort.insertionSortMoveElts(Data, comparator=comparator, lo, hi);
+          return;
+        } else if hi <= lo {
+          // nothing to sort
+          return;
+        }
 
-    var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+        // find pivot using median-of-3 method for small arrays
+        // and a "ninther" for bigger arrays. Places the pivot in
+        // Data[lo].
+        if hi - lo < 100 {
+          piv = order3(Data, lo, mid, hi, comparator);
+        } else {
+          // assumes array size > 9 at the very least
+
+          // median of each group of 3
+          const medLo  = order3(Data, lo,    lo+1, lo+2,  comparator);
+          const medMid = order3(Data, mid-1, mid,  mid+1, comparator);
+          const medHi  = order3(Data, hi-2,  hi-1, hi,    comparator);
+          // median of the medians
+          piv = order3(Data, medLo, medMid, medHi, comparator);
+        }
 
-    if hi-lo < 300 {
-      // stay sequential
-      quickSortImpl(Data, minlen, comparator, lo, eqStart-1);
-      quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
-    } else {
-      // do the subproblems in parallel
-      forall i in 1..2 {
-        if i == 1 then
-          quickSortImpl(Data, minlen, comparator, lo, eqStart-1);
-        else
-          quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
+        var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+
+        // stay sequential
+        if hi-lo < 300 {
```

Comment:
Now the improvement is limited to code that runs only with 300 or fewer elements. But, the `forall` statement below is just checking to see if there are already enough tasks running to keep the cores busy. It might be reasonable to change this branch to say something like `if hi-lo < 300 || runningEnoughParallelTasks {`. The tricky bit is to figure out what exactly to put for `runningEnoughParallelTasks` / how to compute that (where I know of a not-user-facing way to do that but I'm not sure if it's reasonable to do here). Anyway this could be done in another PR.

Reply:
oh, that will be interesting to deal with, I will try to look out some method if am able to I will open a PR for the same :)

Reply:
Aha, you can call `here.runningTasks()`. See https://chapel-lang.org/docs/master/usingchapel/tasks.html#task-related-quantification-methods-on-locales . You can use something like this code:

https://github.com/chapel-lang/chapel/blob/d690e1b53738157f4f5000df9d7c8d08ad4d05c9/modules/internal/ChapelRange.chpl#L2232-L2240

Do you want to try to do it in this PR so we can try for more performance improvement when we measure it?

Reply:
yeah I will try it out pushing the changes now

Reply:
Not sure although if I implemented it the right way or not 😅